packages/perennial/contracts/multiinvoker/MultiinvokerRollup.sol

```diff
@@ -0,0 +1,348 @@
+            ptr += addrLen;
+
+            addr = _getAddressCacheSafe(addrNonceLookup);
+        }
+
+        return (addr, ptr);
+    }
+
+    /**
+     * @notice checked gets the address in cache mapped to the cache index
+     * @dev there is an issue with the calldata if a txn uses cache before caching address
+     * @param nonce The cache index
+     * @return Address stored at cache index
+     */
+    function _getAddressCacheSafe(uint nonce) private view returns (address addr){
+        addr = addressCache[nonce];
+        if (addr == address(0)) revert MultiInvokerRollupInvalidCalldataError();
+    }
+
+    /**
+     * @notice wraps next length of bytes as UFixed18
+     * @param input Full calldata payload
+     * @param ptr Current index of input to start decoding
+     * @param ptr Current index of input to start decoding
+     * @return Current index of calldata to continue decoding
+     */
+     function _decodeAmountUFixed18(bytes calldata input, uint ptr) private view returns (UFixed18 result, uint) {
+        uint temp;
+        (temp, ptr) = _decodeUint(input, ptr);
+
+        return(UFixed18.wrap(temp), ptr);
+    }
+
+    /**
+     * @notice Unpacks next length of bytes into uint256
+     * @param  input Full calldata payload
+     * @param  ptr Current index of input to start decoding
+     * @return Decoded uint
+     * @return Current index of calldata to continue decoding
+     */
+    function _decodeUint(bytes calldata input, uint ptr) private pure returns (uint result, uint) {
+        uint8 len = _bytesToUint8(input[ptr:ptr+1]);
+        ++ptr;
+
+        result = _bytesToUint256(input[ptr:ptr+len]);
+        ptr += len;
+
+        return (result, ptr);
+    }
+
+    /**
+     * @notice Unpacks next length of bytes as lengths of bytes into array of uint256
+     * @param  input Full calldata payload
+     * @param  ptr Current index of input to start decoding
+     * @param  ProgramIds for CLAIM action 
+     * @return Current index of calldata to continue decoding
+     */
+    function _decodeUintArray(bytes calldata input, uint ptr) private pure returns (uint256[] memory, uint) {
+        uint8 arrayLen = _bytesToUint8(input[ptr:ptr+1]);
+        ++ptr;
+
+        uint256[] memory result = new uint256[](arrayLen);
+
+        for (uint count; count < arrayLen;) {
+            uint currUint;
+
+            (currUint, ptr) = _decodeUint(input, ptr);
+
+            result[count] = currUint;
+
+            ++count;
+        }
+        return (result, ptr);
+    }
+
+    /**
+     * @dev This is called in decodeAccount and decodeProduct which both only pass 20 byte slices 
+     * @notice Unchecked force of 20 bytes into address
+     * @param input The 20 bytes to be converted to address
+     * @return Address representation of `input`
+    */
+    function _bytesToAddress(bytes memory input) private pure returns (address addr) {
+        assembly {
+            addr := mload(add(input, 20))
```

Comment:
question: isn't the input of `mload` the starting positing to read `32` bytes from? So in this logic wouldn't we start reading after the address in memory? Or am I misunderstanding how mload works?